### ISSUE TITLE
Iterator with tmin/tmax and ray shortening

### ISSUE THREAD

**Issue by mdesmedt** [svenstaro/bvh#66]
Currently traverse() and traverse_iterator() visit every node along an infinite ray. The application however has information which could reduce traversal steps:

- Provide a `tmin`/`tmax` with the ray and test ray/AABB distance against this range.
- When the app is looking for the smallest `t` along the ray (common) it can shorten the ray during traversal `tmax=min(tmax,t)`

Implementing an iterator which supports tmin/tmax as well as making these mutable to shorten the ray during traversal would help performance in many apps.

**Comment by svenstaro** [comment]
Is this a dup of #56?

**Comment by mdesmedt** [comment]
> Is this a dup of #56?

I don't believe so. #56 asks for "what is the closest object/s in the BVH to arbitrary points".

This request is mostly for adding ray shortening to the traversal. It could look something like:
```
let iter = bvh.traverse_iterator(ray)
for shape in iter {
    let hitinfo = do_full_intersection(ray, shape);
    if hitinfo.hit {
        iter.shorten_ray(hitinfo.t);
    }
}
```

**Comment by mdesmedt** [comment]
I gave this some thought. I think the only way to safely do this is to pass an `Rc<RefCell<Ray>>` to the iterator, which adds a lot of overhead during traversal. And computing t per box is also extra cycles. So unless we add unsafe code (not a fan) or can think of another approach this is probably not worth it. Some form of SIMD is probably still the best option for perf gain.

**Comment by tavianator** [comment]
> And computing t per box is also extra cycles

That's not really true, the slab method computes t anyway.

**Comment by finnbear** [comment]
As of #128, users can do something like this:
```rust
struct ShorteningRay {
    inner: Ray<f32, _>,
    max_t: Rc<RefCell<f32>>,
}

impl IntersectsAabb for ShorteningRay {
    fn intersects_aabb(&self, aabb: &Aabb<f32, _>) -> bool {
       let (nearest_t, _) = self.inner.intersection_slice_for_aabb(aabb);
       let mut max_t = self.max_t.borrow_mut();
       if nearest_t == -1.0 || nearest_t > *max_t {
            return false;
       }
       *max_t = max_t.min(nearest_t);
       true
   }
}
```
